core/src/main/scala/cromwell/core/callcaching/docker/registryv2/DockerRegistryV2AbstractFlow.scala

```diff
@@ -0,0 +1,285 @@
+package cromwell.core.callcaching.docker.registryv2
+
+import akka.NotUsed
+import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport._
+import akka.http.scaladsl.model.HttpHeader.ParsingResult
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.headers.{Authorization, OAuth2BearerToken}
+import akka.http.scaladsl.unmarshalling.Unmarshal
+import akka.stream._
+import akka.stream.scaladsl.{Flow, GraphDSL, Merge}
+import cromwell.core.callcaching.docker.DockerHashActor._
+import cromwell.core.callcaching.docker.registryv2.DockerRegistryV2AbstractFlow._
+import cromwell.core.callcaching.docker.registryv2.flows.{FlowUtils, HttpFlowWithRetry}
+import cromwell.core.callcaching.docker.{DockerFlow, DockerHashResult, DockerImageIdentifierWithoutHash}
+import spray.json.{JsObject, JsString, _}
+
+import scala.collection.immutable.{Seq => ImmutableSeq}
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.implicitConversions
+import scala.util.{Failure, Success, Try}
+
+object DockerRegistryV2AbstractFlow {
+  type HttpDockerFlow = Flow[(HttpRequest, (DockerHashContext, HttpRequest)), (Try[HttpResponse], (DockerHashContext, HttpRequest)), NotUsed]
+  
+  val DigestHeaderName = "Docker-Content-Digest".toLowerCase
+  val AcceptHeader = HttpHeader.parse("Accept", "application/vnd.docker.distribution.manifest.v2+json") match {
+    case ParsingResult.Ok(header, errors) if errors.isEmpty => header
+    case ParsingResult.Ok(header, errors) =>
+      errors foreach { err => logger.warn(err.formatPretty) }
+      header
+    case ParsingResult.Error(error) => throw new RuntimeException(error.formatPretty)
+  }
+}
+
+/**
+  * Implements logic to build a flow that can retrieve a docker hash 
+  * from a registry server conforming to the docker registry API v2 Specification:
+  * https://docs.docker.com/registry/spec/api/
+  */
+abstract class DockerRegistryV2AbstractFlow(httpClientFlow: HttpDockerFlow)(implicit ec: ExecutionContext, materializer: ActorMaterializer) extends DockerFlow {
+  
+  // Http requests take headers as a scala.collection.immutable.Seq
+  // This implicitly converts type Seq collections to immutable
```

Comment:
One of the many reasons I keep asking people to stop using Seq everywhere, in particular with return types (there's no reason a return type needs to be uber-generic).